File: src/test/java/ar/com/flow/bankaccount/balance/BalanceTest.java
package ar.com.flow.bankaccount.balance;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class BalanceTest {
    @Test
    void givenNegativeAmountWhenCreatePositiveBalanceThenItShouldNotBeCreated() {
        assertThrows(IllegalArgumentException.class, () -> new PositiveBalance("USD", -10));
    }

    @Test
    void givenPositiveAmountWhenCreatePositiveBalanceThenItShouldHavePositiveAmount() {
        var balance = new PositiveBalance("USD", 10);

        assertBalance(balance, "USD", 10);
    }

    @Test
    void givenPositiveAmountWhenCreateNegativeBalanceThenItShouldHaveNegativeAmount() {
        var balance = new NegativeBalance("USD", 10);

        assertBalance(balance, "USD", -10);
    }

    @Test
    void givenNegativeAmountWhenCreateNegativeBalanceThenItShouldHaveNegativeAmount() {
        var balance = new NegativeBalance("USD", -10);

        assertBalance(balance, "USD", -10);
    }

    private void assertBalance(Balance balance, String expectedCurrency, int expectedAmount) {
        assertEquals(expectedCurrency, balance.getCurrency());
        assertEquals(expectedAmount, balance.getAmount());
    }
}

File: src/test/kotlin/ar/com/flow/bankaccount/balance/BalanceTest.kt
package ar.com.flow.bankaccount.balance

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class BalanceTest {
    @Test
    fun givenNegativeAmountWhenCreatePositiveBalanceThenItShouldNotBeCreated() {
        Assertions.assertThrows(IllegalArgumentException::class.java) { PositiveBalance.of("USD", -10) }
    }

    @Test
    fun givenPositiveAmountWhenCreatePositiveBalanceThenItShouldHavePositiveAmount() {
        val balance = PositiveBalance.of("USD", 10)
        assertBalance(balance, "USD", 10)
    }

    @Test
    fun givenPositiveAmountWhenCreateNegativeBalanceThenItShouldHaveNegativeAmount() {
        val balance = NegativeBalance.of("USD", 10)
        assertBalance(balance, "USD", -10)
    }

    @Test
    fun givenNegativeAmountWhenCreateNegativeBalanceThenItShouldHaveNegativeAmount() {
        val balance = NegativeBalance.of("USD", -10)
        assertBalance(balance, "USD", -10)
    }

    private fun assertBalance(balance: Balance, expectedCurrency: String, expectedAmount: Int) {
        Assertions.assertEquals(expectedCurrency, balance.currency)
        Assertions.assertEquals(expectedAmount, balance.amount)
    }
}